### PR TITLE
Feature/roles guard

### PR DESCRIPTION
**Pull request by Parle648**
## Describe your changes

- Created roles guard.

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://codecraftersintership.atlassian.net/browse/SCRUM-9?atlOrigin=eyJpIjoiYmFkYjUwMTdhYTcyNGU5OTkxNzZjNDFlYzZhZGUxMjIiLCJwIjoiaiJ9)

### **General**

- [x] types for input and output params
- [x] no `any`, should be also added to eslint rules
- [x] try/catch for any async function
- [x] no **magic** [numbers](/7d77574ee4bc4e339a68066762e887cb)
- [x] compare only with constants not with strings (instead of user === ‘admin’, better user === roles.admin)
- [x] no ternary operator inside ternary operator (bad example: question ? first: second ? third: fourth)
- [x] avoid similar types with duplicating by generics
- [x] no **console.log** in pr
- [x] .env file should be in .gitignore
- [x] delete commented code if it's not part of documentation
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file
- [x] **constants** and **function** names should be lowercase.
- [x] no dates format in the code (like "dd/MM/yyyy”), move it in global constant variable
- [x] import rules. imports should come in a specific order: node modules, absolute path, relative path
- [x] strings should be in the constant variable. Example: instead of ‘15 3 \* \* _’, const EACH_DAY_15h_15min = ‘15 3 _ \* \*’

### Backend

- [x] swagger for each endpoint. add in the documentation, different types of responses.
      on a successful response (2xx), on unsuccessful response (4xx, 5xx)
- [x] less requests to db, all data should be taken with one query
- [x] **public** in method use only is use function externally
- [x] use ConfigService instead of **process.env**
- [x] use @`@index` decorator for frequently requested data
- [x] use transactions if there is a call chain that mutates data in different tables
- [x] add a set of rules for [nestjs](https://github.com/darraghoriordan/eslint-plugin-nestjs-typed) to the .eslint config file. It helps to prevent bugs and confusing moments
- [x] use [REST API Naming Convention](https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5) for endpoints

  P.S: more information about RESTful API design by [Microsof](https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design)t

- [x] use ‘**UUIDs**` for primary key, ** to prevent Enumeration Attacks**

### CODE REVIEW

File: src/common/guards/roles.guard.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ *
+ * use decorator @SetMetadata() below for applying the list of
+ * required roles for endpoint 
+ *   
+ * @Get()
+   @UseGuards(RolesGuard)
+   @SetMetadata('roles', [Roles.SUPERADMIN])
+ */
```
Comment (lyudad) [inline src/common/guards/roles.guard.ts]: why its needed?

Comment (Parle648) [reply src/common/guards/roles.guard.ts]: this is a brief description of how to apply this guard

Comment (Parle648) [reply src/common/guards/roles.guard.ts]: should I delete this comments?

**Comment by Parle648** [comment]
Fix: guard documentation

File: src/common/guards/roles.guard.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { Injectable, CanActivate, ExecutionContext, UnauthorizedException } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { Request } from 'express';
+import * as jwt from 'jsonwebtoken';
+import { Observable } from 'rxjs';
+
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
```
Comment (lyudad) [inline src/common/guards/roles.guard.ts:9]: empty line between functions

Comment (Parle648) [reply src/common/guards/roles.guard.ts:9]: no problem

**Pull request merged**